pkg/storages/s3/s3reader.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/pkg/errors"
+	"github.com/wal-g/tracelog"
+)
+
+var (
+	DebugLogBufferCounter = 0
+)
+
+type s3Reader struct {
+	lastBody      io.ReadCloser
+	folder        *Folder
+	maxRetries    int
+	retryNum      int
+	objectPath    string
+	storageCursor int64
+	maxRetryDelay time.Duration
+	minRetryDelay time.Duration
+	reconnectId   int
+	logDebugId    string // hash from filename and logDebugId - unique logDebugId used only for debug
+}
+
+func (reader *s3Reader) getObjectRange(from, to int64) (*s3.GetObjectOutput, error) {
+	bytesRange := fmt.Sprintf("bytes=%d-", from)
+	if to != 0 {
+		bytesRange += strconv.Itoa(int(to))
+	}
+	input := &s3.GetObjectInput{
+		Bucket: reader.folder.Bucket,
+		Key:    aws.String(reader.objectPath),
+		Range:  aws.String(bytesRange),
+	}
+	reader.debugLog("GetObject with range %s", bytesRange)
+	return reader.folder.S3API.GetObject(input)
+}
+
+func (reader *s3Reader) Read(p []byte) (n int, err error) {
+	reader.debugLog("Read to buffer [%d] bytes", len(p))
+	reconnect := false
+	if reader.lastBody == nil { // initial connect, if lastBody wasn't provided
+		reconnect = true
+	}
+	for {
+		if reconnect {
+			reconnect = false
+			connErr := reader.reconnect()
+			if connErr != nil {
+				reader.debugLog("reconnect failed %s", connErr)
+				return 0, connErr
+			}
+		}
+
+		n, err = reader.lastBody.Read(p)
+		reader.debugLog("read %d, err %s", n, err)
+		if err != nil && err != io.EOF {
+			reconnect = true
+			continue
+		}
+		reader.storageCursor += int64(n)
+		reader.debugLog("success read")
+		return n, err
+	}
+}
+func (reader *s3Reader) getDebugLogLine(format string, v ...interface{}) string {
+	prefix := fmt.Sprintf("s3Reader [%s] ", reader.logDebugId)
+	message := fmt.Sprintf(format, v...)
+	return prefix + message
+}
+
+func (reader *s3Reader) debugLog(format string, v ...interface{}) {
+	tracelog.DebugLogger.Print(reader.getDebugLogLine(format, v...))
+}
+
+func (reader *s3Reader) reconnect() error {
+	failed := 0
+
+	for {
+		reader.reconnectId++
+		object, err := reader.getObjectRange(reader.storageCursor, 0)
+		if err != nil {
+			failed += 1
+			reader.debugLog("reconnect failed [%d/%d]: %s", failed, reader.maxRetries, err)
+			if failed >= reader.maxRetries {
+				return errors.Wrap(err, reader.getDebugLogLine("Too much reconnecting retries"))
+			}
+			sleepTime := reader.getIncrSleep(failed)
+			reader.debugLog("sleep: %s", sleepTime)
+			time.Sleep(sleepTime)
+			continue
+		}
+		failed = 0
+		if reader.lastBody != nil {
+			err = reader.lastBody.Close()
+			if err != nil {
+				msg := reader.getDebugLogLine("We have problems with closing previous connection")
+				tracelog.DebugLogger.Print(msg)
+				return errors.Wrap(err, msg)
+			}
+		}
+		reader.lastBody = object.Body
+		reader.debugLog("reconnect #%d succeeded", reader.reconnectId)
+		break
+	}
+	return nil
+}
+
+func (reader *s3Reader) getIncrSleep(retryCount int) time.Duration {
```

Comment:
maybe add some note regarding name of the implemented algorithm?

Reply:
Added comment, this code was copied form defaultRetryer from s3 package